tests/bluetooth/gatt/src/main.c

```diff
@@ -142,6 +142,157 @@ ZTEST(test_gatt, test_gatt_unregister)
+				       BT_GATT_PERM_READ_AUTHEN | BT_GATT_PERM_WRITE_AUTHEN,
+				       read_test, write_test, test_value),
+	};
+
+	struct bt_gatt_attr local_test1_attrs[] = {
+		/* Vendor Primary Service Declaration */
+		BT_GATT_PRIMARY_SERVICE(&test1_uuid),
+
+		BT_GATT_CHARACTERISTIC(&test1_nfy_uuid.uuid, BT_GATT_CHRC_NOTIFY, BT_GATT_PERM_NONE,
+				       NULL, NULL, &nfy_enabled),
+		BT_GATT_CCC(test1_ccc_cfg_changed, BT_GATT_PERM_READ | BT_GATT_PERM_WRITE),
+	};
+	struct bt_gatt_service local_test_svc = BT_GATT_SERVICE(local_test_attrs);
+	struct bt_gatt_service local_test1_svc = BT_GATT_SERVICE(local_test1_attrs);
+
+	for (int i = 0; i < 10; i++) {
+
+		/* Attempt to register service A */
+		zassert_false(bt_gatt_service_register(&local_test_svc),
+			      "Test service A registration failed");
+
+		/* Attempt to unregister service A */
+		zassert_false(bt_gatt_service_unregister(&local_test_svc),
+			      "Test service A unregister failed");
+
+		/* Attempt to register service B */
+		zassert_false(bt_gatt_service_register(&local_test1_svc),
+			      "Test service B registration failed");
+
+		/* Attempt to re-register service A */
+		zassert_false(bt_gatt_service_register(&local_test_svc),
+			      "Test service A re-registering failed...");
+
+		/* Clean up */
+		zassert_false(bt_gatt_service_unregister(&local_test_svc),
+			      "Test service A unregister failed");
+		zassert_false(bt_gatt_service_unregister(&local_test1_svc),
+			      "Test service B unregister failed");
+	}
+}
+
+ZTEST(test_gatt, test_gatt_reregister_pre_allocated_handles)
+{
+	struct bt_gatt_attr local_test_attrs[] = {
+		/* Vendor Primary Service Declaration */
+		BT_GATT_PRIMARY_SERVICE(&test_uuid),
+
+		BT_GATT_CHARACTERISTIC(&test_chrc_uuid.uuid, BT_GATT_CHRC_READ | BT_GATT_CHRC_WRITE,
+				       BT_GATT_PERM_READ_AUTHEN | BT_GATT_PERM_WRITE_AUTHEN,
+				       read_test, write_test, test_value),
+	};
+
+	struct bt_gatt_attr local_test1_attrs[] = {
+		/* Vendor Primary Service Declaration */
+		BT_GATT_PRIMARY_SERVICE(&test1_uuid),
+
+		BT_GATT_CHARACTERISTIC(&test1_nfy_uuid.uuid, BT_GATT_CHRC_NOTIFY, BT_GATT_PERM_NONE,
+				       NULL, NULL, &nfy_enabled),
+		BT_GATT_CCC(test1_ccc_cfg_changed, BT_GATT_PERM_READ | BT_GATT_PERM_WRITE),
+	};
+	struct bt_gatt_service local_test_svc = BT_GATT_SERVICE(local_test_attrs);
+	struct bt_gatt_service local_test1_svc = BT_GATT_SERVICE(local_test1_attrs);
+
+	/* Pre-allocate handles for both services */
+	for (int j = 0; j < local_test_svc.attr_count; j++) {
+		local_test_svc.attrs[j].handle = 0x0100 + j;
+	}
+	for (int j = 0; j < local_test1_svc.attr_count; j++) {
+		local_test1_svc.attrs[j].handle = 0x0200 + j;
+	}
+
+	/* Attempt to register service A */
+	zassert_false(bt_gatt_service_register(&local_test_svc),
+		      "Test service A registration failed");
+
+	/* Attempt to unregister service A */
+	zassert_false(bt_gatt_service_unregister(&local_test_svc),
```

Comment:
Same comment as above, but here (and below) we should check that the handles are the same after unregistering as they were before registering the service. Otherwise, if it is zero-ed, the service will still be registered, but this won't be what we expect.